File: java/199.binary-tree-right-side-view.java
/*
 * @lc app=leetcode id=199 lang=java
 *
 * [199] Binary Tree Right Side View
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> res = new LinkedList<>();
        if (root == null) return res;
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);

        while (!queue.isEmpty()) {
            int len = queue.size();
            TreeNode node = null;
            for (int i = 0; i < len; ++i) {
                node = queue.poll();
                if (node.left != null) queue.offer(node.left);
                if (node.right != null) queue.offer(node.right);
            }
            res.add(node.val);
        }

        return res;
    }
}

File: go/199.binary-tree-right-side-view.go
/*
 * @lc app=leetcode id=199 lang=golang
 *
 * [199] Binary Tree Right Side View
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func rightSideView(root *TreeNode) []int {
	res := make([]int, 0)
    if root == nil {
		return res
	}
	queue := []*TreeNode{root}
	curcnt, nextcnt := 1, 0
	for len(queue) != 0 {
		for i := 0; i < curcnt; i++ {
			node := queue[i]
			if node.Left != nil {
				queue = append(queue, node.Left)
				nextcnt++
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
				nextcnt++
			}
		}
		res = append(res, queue[curcnt-1].Val)
		queue = queue[curcnt:]
		curcnt, nextcnt = nextcnt, 0
	}
	return res
}
// @lc code=end

